zproject/backends.py

```diff
@@ -550,11 +563,91 @@ def get_authenticated_user(self, *args: Any, **kwargs: Any) -> Optional[UserProf
 
         return user_profile
 
+class SAMLAuthBackend(SocialAuthMixin, SAMLAuth):
+    auth_backend_name = "SAML"
+
+    def auth_complete(self, *args, **kwargs):
+        """
+        The user has been redirected back from the IdP and we should
+        now log them in, if everything checks out.
+        """
+        # This is copy-pasted from the SAML backend to use our own do_auth
+        # instead of self.strategy.authenticate
+        idp_name = self.strategy.request_data()['RelayState']
+        idp = self.get_idp(idp_name)
+        auth = self._create_saml_auth(idp)
+        auth.process_response()
+        errors = auth.get_errors()
+        if errors or not auth.is_authenticated():
+            reason = auth.get_last_error_reason()
+            raise AuthFailed(
+                self, 'SAML login failed: {0} ({1})'.format(errors, reason)
+            )
+
+        attributes = auth.get_attributes()
+        attributes['name_id'] = auth.get_nameid()
+        self._check_entitlements(idp, attributes)
+        response = {
+            'idp_name': idp_name,
+            'attributes': attributes,
+            'session_index': auth.get_session_index(),
+        }
+        kwargs.update({'response': response, 'backend': self})
+        return self.do_auth(*args, **kwargs)
+
+    def get_email_address(self, *args: Any, **kwargs: Any) -> Optional[str]:
+        try:
+            return self.get_user_details(kwargs['response'])['email']
+        except KeyError:  # nocoverage # TODO: investigate
+            return None
```

Comment:
I think you can just remove this function (as well as `get_full_name`) ; it appears `get_user_details` in the `SAMLIdentityProvider` base class in python-social-auth set the email properly, so the common code we have for fetching this from the `details` dict should Just Work.